Initial attempt at documenting new ODE interface (design-doc #19) (#213)
stan-dev/docs: src/stan-users-guide/odes.Rmd

  [@CohenHindmarsh:1996; @SerbanHindmarsh:2005]

For a discussion of stiff ODE systems, see the [stiff ODE

> section](#stiff-ode.section). The function signatures for Stan's ODE

Author: > This is due to the attempt of Adams to do a higher order approximation of the function in order to make large steps and in turn avoid excessive calls to the ODE RHS (whereas RK45 just calls brutally the ODE RHS very often).

I think we should work on a separate workflow ODE document because we have no real guidance on picking between the solvers. I'd like to just leave this for the future.

> section](#stiff-ode.section). The function signatures for Stan's ODE

Reviewer: I find the text as is confusing as there is no guidance at all to pick between the two non-stiff ones. I would prefer to tell people: Start with RK45 in any case; if stiffness is a concern go with bdf and whenever the problem is non-stiff and the ODE RHS is expensive to evaluate, then try Adams (after having tried RK45).

This would reflect my intuition and I think is a fair summary.  If you think we are not ready for this... ok.

## Coding an ODE System

The simple harmonic oscillator can be coded using the following function
in Stan (see the [user-defined functions chapter](#functions-programming.chapter) for
more information on coding user-defined functions).

```
vector sho(real t,        // time
           vector y,      // state
           real theta) {
  vector[2] dydt;
  dydt[1] = y[2];
  dydt[2] = -y[1] - theta * y[2];
  return dydt;

> }

Author: I agree style-wise, but right now:

```
return [...]';
```

is a little slow. See https://github.com/stan-dev/stanc3/issues/578.

## Coding an ODE System

The simple harmonic oscillator can be coded using the following function
in Stan (see the [user-defined functions chapter](#functions-programming.chapter) for
more information on coding user-defined functions).

```
vector sho(real t,        // time
           vector y,      // state
           real theta) {
  vector[2] dydt;
  dydt[1] = y[2];
  dydt[2] = -y[1] - theta * y[2];
  return dydt;
}
```

The function takes in a time `t` (a real value), the system state
`y` (a vector), and the parameter theta. The function returns a vector
of time derivatives of the system state at time `t`, state `y`, and parameter
`theta`. The simple harmonic oscillator coded here does not have time-sensitive
equations; that is, `t` does not show up in the definition of
`dydt`, however it is still required.

### Strict Signature {-}

> Every ODE system function has the same first two arguments and the
> same return type, but the third argument onwards may change be anything.

Reviewer 2: ```suggestion
Every ODE system function must have the first two arguments be a `real` and `vector` type, but ODE functions may have any number and type of arguments as the rest of the function signature.
```

For instance, all of these are valid ODE signatures:

```
vector myode1(real t, vector y, real a0);
vector myode2(real t, vector y, int[] a0, vector a1);
vector myode3(real t, vector y, matrix a0, real[] a1, row_vector a2);
```

> And these are not allowed:

Reviewer 2: ```suggestion
But these are not allowed:
```

```
functions {

>   vector sho(real t,

Reviewer 3: Yes.  You can use stan mode.  Whether the resulting syntax highlighting is "nice" is another matter.

>   vector sho(real t,

Author: We aren't using this elsewhere in the manual, so I'm going to leave it off here.

using the matrix exponential function, as the method can, depending
on the system at hand, yield more efficient results.

This program illustrates the way in which the ODE solver is called in
a Stan program,

```
y_hat = integrate_ode_rk45(sho, y0, t0, ts, theta, x_r, x_i);
vector[2] y_hat[T] = ode_rk45(sho, y0, t0, ts, theta);
```

This assigns the solutions to the system defined by function
`sho`, given initial state `y0`, initial time `t0`,
requested solution times `ts`, parameters `theta`, real data
`x`, and integer data `x_int`.  The call explicitly
this returns the solution of the ODE initial value problem defined
by system function `sho`, initial state `y0`, initial time `t0`, and
parameter `theta` at the times `ts`. The call explicitly
specifies the Runge-Kutta solver (for non-stiff systems).

Here, the ODE solver is called in the generated quantities block to
provide a $10 \times 2$ array of solutions `y_hat` to
which measurement error is added using the normal pseudo-random number
generating function `normal_rng`.  The number of rows in the
solution array is the same as the size of `ts`, the requested
solution times.

> ### Data versus Parameters {-}

Author: I removed this because the new interface doesn't have this limitation (it can take any combination of data/param arguments).

In the new interface you might accidentally over-promote a data thing by defining it in transformed parameters, but whether it is defined in transformed parameters or transformed data the solver will still work.

             real theta) {
    vector[2] dydt;
    dydt[1] = y[2];
    dydt[2] = -y[1] - theta * y[2];
    return dydt;
  }
}
data {
  int<lower=1> T;
  vector[2] y[T];
  real t0;
  real ts[T];
}
parameters {
  real y0[2];
  vector<lower=0>[2] sigma;
  real theta;
}
model {
  vector[2] y_hat[T];
  sigma ~ normal(0, 2.5);
  theta ~ std_normal();
  y0 ~ std_normal();
  y_hat = ode_rk45(sho, y0, t0, ts, theta);
  for (t in 1:T)
    y[t] ~ normal(y_hat[t], sigma);
}
```
<a name="id:sho-both.figure"></a>

Compared to the program for simulation,
the program to estimate parameters uses the `ode_rk45` function
in the model block rather than the generated quantities block.  There
are half-normal priors on the measurement error scales `sigma` and standard
normal priors on the components of parameter array `theta` and initial
state parameter array `y0`.  The solutions to the ODE are
assigned to `y_hat`, which is then used as the location in
the observation noise model as follows.

```
vector[2] y_hat[T] = ode_rk45(sho, y0, t0, ts, theta);
for (t in 1:T)

>   y[t] ~ normal(y_hat[t], sigma);

Author: This should be efficient. The ODE solve only happens once outside the for loop.

The elements of `y` are vectors so the sampling statement is already vectorized here.

using the matrix exponential function, as the method can, depending
on the system at hand, yield more efficient results.

This program illustrates the way in which the ODE solver is called in
a Stan program,

```
vector[2] y_hat[T] = ode_rk45(sho, y0, t0, ts, theta);
```

this returns the solution of the ODE initial value problem defined
by system function `sho`, initial state `y0`, initial time `t0`, and
parameter `theta` at the times `ts`. The call explicitly
specifies the Runge-Kutta solver (for non-stiff systems).

Here, the ODE solver is called in the generated quantities block to
compute the solution of the ODE at `T` output times and add simulated
measurement noise using `normal_rng`.

### Estimating System Parameters and Initial State {-}

> With noisy observations of the solution it is possible in Stan to estimate
> both parameters and initial conditions. As an example, this is a harmonic
> oscillator example where both the system parameter `theta` and initial state
> `y0` are parameters:

Reviewer 2: ```suggestion
When the solution's observations are noisy it is possible in Stan to estimate
both the parameters and initial conditions. The example below is a harmonic
oscillator where both the system parameter `theta` and initial state
`y0` are parameters:
```

```
functions {
  vector sho(real t,
             vector y,
             real theta) {
    vector[2] dydt;
    dydt[1] = y[2];
    dydt[2] = -y[1] - theta * y[2];
    return dydt;
  }
}
data {
  int<lower=1> T;
  vector[2] y[T];
  real t0;
  real ts[T];
}
parameters {
  real y0[2];
  vector<lower=0>[2] sigma;
  real theta;
}
model {
  vector[2] y_hat[T];
  sigma ~ normal(0, 2.5);
  theta ~ std_normal();
  y0 ~ std_normal();
  y_hat = ode_rk45(sho, y0, t0, ts, theta);
  for (t in 1:T)
    y[t] ~ normal(y_hat[t], sigma);
}
```
<a name="id:sho-both.figure"></a>

Compared to the program for simulation,
the program to estimate parameters uses the `ode_rk45` function
in the model block rather than the generated quantities block.  There
are half-normal priors on the measurement error scales `sigma` and standard
normal priors on the components of parameter array `theta` and initial
state parameter array `y0`.  The solutions to the ODE are

> assigned to `y_hat`, which is then used as the location in

Reviewer 2: Is this example in the example repo as well? If so may be good to update there as well

Author: I checked and the footnotes stuff hasn't made it this far in the manual yet, so there's no updating to be done there.

Reviewer 3: @SteveBronder was talking about the `example-models` repo.   

The first footnotes PR is still pending:  https://github.com/stan-dev/docs/pull/164

the observation noise model as follows.

```
vector[2] y_hat[T] = ode_rk45(sho, y0, t0, ts, theta);
for (t in 1:T)
  y[t] ~ normal(y_hat[t], sigma);
```

As with other regression-like models, it's easy to change the noise

> model to something with heavier tails (e.g., Student-t distributed),
> to be correlated in the state variables (e.g., with a multivariate
> normal distribution), or both (e.g., with a multivariate Student-t
> distribution).

Reviewer 2: ```suggestion
model to something with heavier tails (e.g., Student-t distributed),
correlation in the state variables (e.g., with a multivariate
normal distribution), or both heavy tails and correlation in the state variables (e.g., with a multivariate Student-t
distribution).
```

with a stiff solver. On the other hand, attempting to use the non-stiff
integrator for a stiff system will cause the timestep to become very small,
leading the non-stiff solver taking more time overall even if each timestep
is easier to compute than for the stiff solver.

## Control Parameters for ODE Solving

The calls to the integrators shown above just used the default
control settings.  Both the non-stiff and stiff integrators have an additional
signature that makes it possible to specify the `relative_tolerance`,
`absolute_tolerance`, and `max_num_steps` control parameters.
All three additional arguments must be supplied if any of them is required.

```
vector[2] y_hat[T] = ode_bdf_tol(sho, y0, t0, ts,
                                 relative_tolerance,
				                         absolute_tolerance,
				                         max_num_steps,
				                         theta);
```

`relative_tolerance` and `absolute_tolerance` control the stepsize and
`max_num_steps` specifies the maximum number of steps the solver will
take between output time points before giving up.

For the RK45 solver, the default values for relative and absolute tolerance are

> both $10^{-6}$ and the maximum number of steps between outputs is

Author: Lol, symmetry is fun. I'm leaving this out though.

one million. For the BDF and Adams solvers, the relative and absolute
tolerances are $10^{-10}$ and the maximum number of steps between outputs is

> one hundred million.

Author: I just took this default from the old code. I think you're right. I made an issue with computational motivation to make this smaller for both solvers: https://github.com/stan-dev/math/issues/1969

> one hundred million.

Author: (But I'm going to leave this as it is for now)

### Discontinuous ODE System Function {-}

If there are discontinuities in the ODE system function, it is best

> to integrate the ODE to those discontinuities, stop the solver, and

Reviewer 3: "to those" -> "between the"

restart it on the other side.

Nonetheless, the ODE integrator will attempt to integrate over discontinuities
in the state function. The accuracy of points near the discontinuity
may be problematic (requiring many small steps).  An example of such a
discontinuity is a lag in a pharmacokinetic model, where a
concentration is going to be zero for times $0 < t < t'$ for some
lag-time $t'$, whereas it will be nonzero for times $t \geq t'$.
In this example example, we would use code in the system such as

```
if (t < t_lag)
  return 0;
else
  ... return non-zero value...;
```

In general it is better to integrate up to `t_lag` in one solve and

> then integrate from `t_lag` onwards in another.

Reviewer 3: Cool.  I was wondering about that.  Definitely include this.

> then integrate from `t_lag` onwards in another.

Author: > then he should at least output a solution in the vicinity or at the location of the discontinuity a value from the ODE solver. This forces the ODE solver to actually "look" at the time point in question as to what happens

I thought the ODE solvers interpolate to get their output points (and so the actual steps the solvers take are separate). I don't think adding output points will improve the situation.

both $10^{-6}$ and the maximum number of steps between outputs is
one million. For the BDF and Adams solvers, the relative and absolute
tolerances are $10^{-10}$ and the maximum number of steps between outputs is
one hundred million.

### Discontinuous ODE System Function {-}

If there are discontinuities in the ODE system function, it is best
to integrate the ODE between the discontinuities, stop the solver at each
one, and restart it on the other side.

Nonetheless, the ODE integrator will attempt to integrate over discontinuities
it encounters in the state function. The accuracy of the solution near the
discontinuity may be problematic (requiring many small steps). An example of
such a discontinuity is a lag in a pharmacokinetic model, where a
concentration is zero for times $0 < t < t'$ and then positive for $t \geq t'$.
In this example example, we would use code in the system such as

```
if (t < t_lag)
  return 0;
else
  ... return non-zero value...;
```

In general it is better to integrate up to `t_lag` in one solve and
then integrate from `t_lag` onwards in another. Mathematically, the
discontinuity can make the problem ill-defined and the numerical integrator

> may behave erratically around it.

Author: > then it's strongly advisable to output some function values in the vicinity of the discontinuity

I don't think this will improve the behavior of the integrator around those points though. Intermediate outputs are just interpolated from the actual path the solver took.

I guess I could suggest this if there was something to do with the output. Like make plots or something. I just don't think there's an obvious way to diagnose if there are problems or not here.

> may behave erratically around it.

Reviewer: No, it really does make a difference. I had experienced this myself when coding dosing events as approximate delta peaks. Only when I added observation time points around the dosing time, only then the integrator picked up the dosing event. The big disadvantage of the approach you propose to restart is that you must restart the integration process with the initials being parameters whereas if you keep integrating in a single go, then - for pharmacokinetic problems at least - you can use data as the initial. This will dramatically speedup the solve process as the sensitivity system is a lot smaller as compared to making initials varying.

as in the previous example code.

In general, computing a matrix exponential will be more efficient than using a numerical
solver. We can however only apply this technique to systems of linear ODEs.

```
data {
  int<lower=1> T;
  vector[2] y0;
  real ts[T];
  real theta[1];
}
model {
}
generated quantities {
  vector[2] y_hat[T];
  matrix[2, 2] A = [[ 0,  1],
                    [-1, -theta[1]]]
  for (t in 1:T)
    y_hat[t] = matrix_exp((t - 1) * A) * y0;
  // add measurement error
  for (t in 1:T) {
    y_hat[t, 1] += normal_rng(0, 0.1);
    y_hat[t, 2] += normal_rng(0, 0.1);
  }
}
```
<a name="id:sho-sim-me.figure"></a>

This Stan program simulates noisy measurements from a simple harmonic
oscillator.  The system of linear differential equations is coded as a
matrix. The system parameters `theta` and initial state `y0` are read
in as data along observation times `ts`.  The generated quantities
block is used to solve the ODE for the specified times and then add
random measurement error, producing observations `y_hat`. Because the
ODEs are linear, we can use the `matrix_exp` function to solve the
system.

> 

Reviewer 3: Is there a way to get a steady state solution to a linear system using the matrix exponential?

Author: I think you'd do that by saying `y' = 0` for steady state means `0 = A * y` and them solving for `y`, which I guess is zero if `A` is full rank? I'm not sure but I don't think you'd do the matrix exponential.

#### Sensitivity Analysis {-}

> The tolerances should be set low enough that setting them lower does

Reviewer 3: This is a confusing title.  This is not sensitivity analysis of the ODE, but of the ODE solver.  Can we call this something like "computational methodology" or "debugging" or "advice on setting tolerances" or something like that?

Author: This makes more sense in the tolerance section so I moved the comment there.
